### ISSUE TITLE
Integer literals 'mis'-interpretation

### ISSUE THREAD

**Issue by jsmall-nvidia** [shader-slang/slang#1043]
Take the shader 

```
struct PSInput
{
	float4 color : COLOR;
};

float4 PSMain(PSInput input) : SV_TARGET
{
    int a = 0xffffffff;
	return input.color * a;
}
```

a (assuming int is 32bits as it is in hlsl) is the value -1. Dxc recognizes this and outputs in effect `return -input.color`. Slang does not and outputs...

```
[numthreads(1, 1, 1)]
vector<float,4> PSMain(PSInput_0 input_0) : SV_TARGET
{
    return input_0.color_0 * (float) 4294967295;
}
```

Note that if we do `int a = int(0xffffffff);` the cast appears in the output and so does produce something that should produce the same result.

```
[numthreads(1, 1, 1)]
vector<float,4> PSMain(PSInput_0 input_0) : SV_TARGET
{
    return input_0.color_0 * (float) ((int) 4294967295);
}
```

**Comment by tfoleyNV** [comment]
It seems like the underlying problem is that we might have an integer literal with a known type in the IR, but when we output it in generated source we don't include that type anywhere.

For example, both of the following would seemingly generate the same output:

```hlsl
// using 32-bit type
int a = 0xffffffff;
return double(1.0) * a;

// using 64-bit type
int64_t a = 0xffffffff;
return double(1.0) * a;
```
In both cases the output today would be `double(1.0) * 4292967295`.

The best solution is probably to ensure that literals (or at least *integer* literals) are always emitted with some kind of explicit type ascription (whether via a suffix or a cast). If we always emit `int(...)` or `int64_t(...)` for a literal then this should become a non-issue.

**Comment by jsmall-nvidia** [comment]
This was addressed in #1202

The short version is that now literals are sign-extended/converted such they fit within their literal type. Warning are given for loss of bits in common cases. 

It is arguable whether this should be the final solution. Other mechanisms are discussed in issue #1185.
